bip-musig2.mediawiki

```diff
@@ -0,0 +1,761 @@
+The security of the resulting scheme is then depending on the requirement that reading the counter must never yield the same counter value in two ''NonceGen'' invocations with the same ''sk''.
+
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that the only optional argument is ''rand'', which can be omitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''aggothernonce'' which should be set to the output of ''NonceAgg'' run on the ''pubnonce'' value of '''all''' other signers (but can be provided by an untrusted party).
+Hence, using ''DeterministicSign'' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
+
+==== Deterministic and Stateless Signing for a Single Signer ====
+
+<div>
+Algorithm ''DeterministicSign(sk, aggothernonce, pk<sub>1..u</sub>, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m, rand)'':
+* Inputs:
+** The secret signing key ''sk'': a 32-byte array
+** The aggregate public nonce ''aggothernonce'' (see [[#modifications-to-nonce-generation|above]]): a 66-byte array
+** The number ''u'' of public keys with ''0 < u < 2^32''
+** The plain public keys ''pk<sub>1..u</sub>'': ''u'' 32-byte arrays
+** The number ''v'' of tweaks with ''0 &le; v < 2^32''
+** The tweaks ''tweak<sub>1..v</sub>'': ''v'' 32-byte arrays
+** The tweak methods ''is_xonly_t<sub>1..v</sub>'': ''v'' booleans
+** The message ''m'': a byte array<ref name="mlen" />
+** The auxiliary randomness ''rand'': a 32-byte array (optional argument)
+* If the optional argument ''rand'' is present:
+** Let ''sk' '' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand)''
+* Else:
+** Let ''sk' = sk''
+* Let ''keygen_ctx<sub>0</sub> = KeyAgg(pk<sub>1..u</sub>)''; fail if that fails
+* For ''i = 1 .. v'':
+** Let ''keygen_ctx<sub>i</sub> = ApplyTweak(keygen_ctx<sub>i-1</sub>, tweak<sub>i</sub>, is_xonly_t<sub>i</sub>)''; fail if that fails
+* Let ''aggpk = GetPubkey(keygen_ctx<sub>v</sub>)''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/deterministic/nonce</sub>(sk' || aggothernonce || aggpk || bytes(8, len(m)) || m || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* Fail if ''k<sub>1</sub> = 0'' or ''k<sub>2</sub> = 0''
+* Let ''R<sub>⁎,1</sub> = k<sub>1</sub>⋅G, R<sub>⁎,2</sub> = k<sub>2</sub>⋅G''
+* Let ''pubnonce = cbytes(R<sub>⁎,2</sub>) || cbytes(R<sub>⁎,2</sub>)''
+* Let ''d = int(sk)''
+* Fail if ''d = 0'' or ''d &ge; n''
+* Let ''pk = cbytes(d⋅G)''
+* Let ''secnonce = bytes(32, k<sub>1</sub>) || bytes(32, k<sub>2</sub>) || pk''
+* Let ''aggnonce = NonceAgg((pubnonce, aggothernonce))''; fail if that fails and blame nonce aggregator for invalid ''aggothernonce''.
```

Comment:
The reason that there's no independent `DeterministicNonceGen` is that it is extremely fragile. When it is not used correctly, the secret signing key can leak to the public. For example, `aggnonce` must be the result of calling `NonceAgg` with the deterministic signers nonce and `aggothernonce`, which must be input to the nonce hash. If we had an independent `DeterministicNonceGen` instead of a `DeterministicSign` algorithm, we'd have to describe all of this in prose, which is susceptible to accidental misuse. `DeterministicSign`, on the other hand, is safe to use and can be tested using the provided test vectors.

Yes, it's possible to securely modify the stateless signer to require partial signatures from other parties before creating and publishing its partial signature. For example, this would be achieved by the following two algorithms wrapping `deterministic_sign`.

```python
def deterministic_sign_stage1(sk, aggothernonce, pubkeys, tweaks, is_xonly, msg, rand):
  pubnonce, _ = deterministic_sign(sk, aggothernonce, pubkeys, tweaks, is_xonly, msg, rand)
  return pubnonce

def deterministic_sign_stage2(sk, aggothernonce, pubkeys, tweaks, is_xonly, msg, rand, psigs):
  _, psig = deterministic_sign(sk, aggothernonce, pubkeys, tweaks, is_xonly, msg, rand)
  for i in range(len(pubkeys)):
    if pubkeys[i] == plain_pk_gen(sk):
      continue
    assert partial_sig_verify(psigs[i], pubnonces, pubkeys, tweaks, is_xonly, msg, i)
  return psig
```

Note that the `stage2` requires recomputing a bunch of things that have been computed in `stage1` already but this is necessary to keep the signer stateless. Similarly, if the inputs in stage2 differ from `stage1`, the signer may produce a `psig` that doesn't match the pubnonce. It's possible in `stage1` to call a modified `deterministic_sign` that aborts as soon as the `pubnonce` is computed and does not compute a partial signature that is thrown away anyway. This is easy to see since the observable behavior of that `stage1` is indistinguishable from the original one.